### PR TITLE
SlackITAdmin

### PR DESCRIPTION
**Pull request by sonaaru**
## XSOAR [Partner](https://xsoar.pan.dev/docs/partners/become-a-tech-partner)?
PANW IT Team

## Status
- [ ] In Progress
- [X] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
N/A. New Integration.

## Description
Initial commit.

## Screenshots
N/A

## Minimum version of Demisto
- [X] 5.0.0
- [ ] 5.5.0
- [ ] 6.0.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [X] No

## Must have
- [X] Tests
- [X] Documentation

### CODE REVIEW

File: Packs/SlackITAdmin/Integrations/Slack_IT_Admin/Slack_IT_Admin.py [hunk 1]
```diff
@@ -0,0 +1,724 @@
...
+
+''' HELPER FUNCTIONS '''
+
+
+def map_scim(clientData):
+    try:
+        clientData = json.loads(clientData)
+    except Exception:
+        pass
+    if type(clientData) != dict:
```
Comment (dantavori) [inline Packs/SlackITAdmin/Integrations/Slack_IT_Admin/Slack_IT_Admin.py:221]:
```suggestion
def map_scim(client_data):
    try:
        client_data = json.loads(client_data)
    except Exception:
        pass
    if type(client_data) != dict:
```

File: Packs/SlackITAdmin/Integrations/Slack_IT_Admin/Slack_IT_Admin.py [hunk 2]
```diff
@@ -0,0 +1,724 @@
...
+        "thumbnnail": "photos(val.type && val.type=='thumbnail').[0].value",
+        "timezone": "timezone",
+        "title": "title",
+        "userName": "userName",
+        "userType": "userType",
+    }
+    ret = dict()
+    for k, v in mapping.items():
+        try:
+            ret[k] = demisto.dt(clientData, v)
```
Comment (dantavori) [inline Packs/SlackITAdmin/Integrations/Slack_IT_Admin/Slack_IT_Admin.py:267]:
```suggestion
            ret[k] = demisto.dt(client_data, v)
```

File: Packs/SlackITAdmin/Integrations/Slack_IT_Admin/Slack_IT_Admin.py [hunk 3]
```diff
@@ -0,0 +1,724 @@
...
+        "password": "password",
+        "photo": "photos(val.type && val.type=='photo').[0].value",
+        "preferredLanguage": "preferredLanguage",
+        "profileUrl": "profileUrl",
+        "thumbnnail": "photos(val.type && val.type=='thumbnail').[0].value",
+        "timezone": "timezone",
+        "title": "title",
+        "userName": "userName",
+        "userType": "userType",
+    }
```
Comment (dantavori) [inline Packs/SlackITAdmin/Integrations/Slack_IT_Admin/Slack_IT_Admin.py:263]: do we really need all this? see when map_scim() is used in the integration and what are what are we doing with its returned value

File: Packs/SlackITAdmin/Integrations/Slack_IT_Admin/Slack_IT_Admin.py [hunk 4]
```diff
@@ -0,0 +1,724 @@
...
+
+    if res.status_code != 204:
+        error_json = res.json()
+        return_error(f"Error in API call. Status Code: [{res.status_code}]. Error Response: {error_json}")
+
+    readable_output = f"Updated Slack Group Members for group : {group_id}"
+
+    return (
+        readable_output,
+        {},
```
Comment (dantavori) [inline Packs/SlackITAdmin/Integrations/Slack_IT_Admin/Slack_IT_Admin.py:632]: why isn't anything returned here? (as opposed to what's described in the yml)

File: Packs/SlackITAdmin/Integrations/Slack_IT_Admin/Slack_IT_Admin.py [hunk 5]
```diff
@@ -0,0 +1,724 @@
...
+    res = client.get_group_by_id(group_id)
+    res_json = res.json()
+
+    if res.status_code != 200:
+        return_error(f"Error in API call. Status Code: [{res.status_code}]. Error Response: {res_json}")
+
+    members = res_json.get('members')
+    outputs = {
+        'Slack.Group(val.Id && val.Id === obj.Id)': {
+            'Id': group_id,
```
Comment (dantavori) [inline Packs/SlackITAdmin/Integrations/Slack_IT_Admin/Slack_IT_Admin.py:521]: what if group id wasn't supplied?

please check what is supplied - id or name, and accordingly decide which field should be added. If a name was supplied, make sure the key of the output object will be `Slack.Group(val.Name && val.Name === obj.Name)`.

also make sure you are consistent with other commands and with the yml file regarding this issue.

File: Packs/SlackITAdmin/Integrations/Slack_IT_Admin/Slack_IT_Admin.py [hunk 6]
```diff
@@ -0,0 +1,724 @@
...
+    res = client.delete_group(group_id)
+
+    if res.status_code != 204:
+        error_json = res.json()
+        return_error(f"Error in API call. Status Code: [{res.status_code}]. Error Response: {error_json}")
+
+    readable_output = f'Slack Group: "{group_id}" was deleted successfully'
+    return (
+        readable_output,
+        {},
```
Comment (dantavori) [inline Packs/SlackITAdmin/Integrations/Slack_IT_Admin/Slack_IT_Admin.py:555]: please add to the context the group data (which in this case will be the group name/id) `"deleted": True` (and for other commands `"deleted": False`) and update in the context output in the yml as well.
The output object should look similar to what I described in my comment in get_group() method.

File: Packs/SlackITAdmin/Integrations/Slack_IT_Admin/Slack_IT_Admin.py [hunk 7]
```diff
@@ -0,0 +1,724 @@
...
+    res = client.create_group(group_data)
+    res_json = res.json()
+
+    if res.status_code not in [200, 201]:
+        return_error(f"Error in API call. Status Code: [{res.status_code}]. Error Response: {res_json}")
+
+    readable_output = f'Slack Group "{res_json.get("id")}" created successfully'
+    return (
+        readable_output,
+        {},
```
Comment (dantavori) [inline Packs/SlackITAdmin/Integrations/Slack_IT_Admin/Slack_IT_Admin.py:583]: please add the data to the context as well (and update in the yml the context output)

File: Packs/SlackITAdmin/Integrations/Slack_IT_Admin/Slack_IT_Admin.py [hunk 8]
```diff
@@ -0,0 +1,724 @@
...
+
+    if res.status_code not in [200, 201]:
+        return_error(f"Error in API call. Status Code: [{res.status_code}]. Error Response: {res_json}")
+
+    readable_output = tableToMarkdown(name=f"Slack Group replaced: {group_id}",
+                                      t=res_json)
+
+    return (
+        readable_output,
+        {},
```
Comment (dantavori) [inline Packs/SlackITAdmin/Integrations/Slack_IT_Admin/Slack_IT_Admin.py:671]: please add the data to the context as well (and update in the yml the context output)

**Review by dantavori — CHANGES_REQUESTED** [review]
see my notes - also when you are done, regenerate the readme file because it needs to be updated according to the changes in the yml.

File: Packs/SlackITAdmin/Integrations/Slack_IT_Admin/Slack_IT_Admin.py [hunk 9]
```diff
@@ -0,0 +1,724 @@
...
+        res_json)
+
+
+def main():
+    """
+        PARSE AND VALIDATE INTEGRATION PARAMS
+    """
+    params = demisto.params()
+    # get the service API url
+    base_url = 'https://api.slack.com/scim/v1/'
```
Comment (dantavori) [inline Packs/SlackITAdmin/Integrations/Slack_IT_Admin/Slack_IT_Admin.py:681]: please make it a global variable at the top of the script

File: Packs/SlackITAdmin/Integrations/Slack_IT_Admin/Slack_IT_Admin.py [hunk 10]
```diff
@@ -0,0 +1,724 @@
...
+    if not group_id:
+        group_id = client.get_group_id(group_name)
+
+    res = client.delete_group(group_id)
+
+    if res.status_code != 204:
+        error_json = res.json()
+        return_error(f"Error in API call. Status Code: [{res.status_code}]. Error Response: {error_json}")
+
+    readable_output = f'Slack Group: "{group_id}" was deleted successfully'
```
[the remark this answers is not in the document]
Comment (sonaaru) [reply Packs/SlackITAdmin/Integrations/Slack_IT_Admin/Slack_IT_Admin.py:552]: code expects either groupId or groupName. If user passes groupName, group_id will be fetched from client.get_group_id method (by API call) and then API will be called to delete group. So variable "group_id" will have value always in this line.

File: Packs/SlackITAdmin/Integrations/Slack_IT_Admin/Slack_IT_Admin.py [hunk 11]
```diff
@@ -0,0 +1,724 @@
...
+    members = res_json.get('members')
+    outputs = {
+        'Slack.Group(val.Id && val.Id === obj.Id)': {
+            'Id': group_id,
+            'DisplayName': res_json.get('displayName'),
+            'Members': members
+        }
+    }
+
+    readable_output = tableToMarkdown(name=f"Slack Group Members: {res_json.get('displayName')} [{group_id}]",
```
[the remark this answers is not in the document]
Comment (sonaaru) [reply Packs/SlackITAdmin/Integrations/Slack_IT_Admin/Slack_IT_Admin.py:527]: code expects either groupId or groupName. If user passes groupName, group_id will be fetched from client.get_group_id method (by API call) and then API will be called to delete group. So variable "group_id" will have value always in this line.

File: Packs/SlackITAdmin/Integrations/Slack_IT_Admin/Slack_IT_Admin.py (hunk 5 above)
Comment (sonaaru) [reply Packs/SlackITAdmin/Integrations/Slack_IT_Admin/Slack_IT_Admin.py:519]: code expects either groupId or groupName. If user passes groupName, group_id will be fetched from client.get_group_id method (by API call) and then API will be called to delete group. So variable "group_id" will have value always in this line.

File: Packs/SlackITAdmin/Integrations/Slack_IT_Admin/Slack_IT_Admin.py [hunk 12]
```diff
@@ -0,0 +1,724 @@
...
+    Client will implement the service API, and should not contain any Demisto logic.
+    Should only do requests and return data.
+    """
+    def __init__(self, base_url, verify=True, proxy=False, headers=None):
+        self.base_url = base_url
+        self.verify = verify
+        self.headers = headers
+        self.session = requests.Session()
+        if not proxy:
+            self.session.trust_env = False
```
[the remark this answers is not in the document]
Comment (sonaaru) [reply Packs/SlackITAdmin/Integrations/Slack_IT_Admin/Slack_IT_Admin.py]: Removed.

File: Packs/SlackITAdmin/Integrations/Slack_IT_Admin/Slack_IT_Admin.py [hunk 13]
```diff
@@ -0,0 +1,724 @@
...
+
+    return (
+        readable_output,
+        {},
+        None
+    )
+
+
+def replace_group_command(client, args):
+    demisto.log(str(args))
```
[the remark this answers is not in the document]
Comment (sonaaru) [reply Packs/SlackITAdmin/Integrations/Slack_IT_Admin/Slack_IT_Admin.py]: Removed.

File: Packs/SlackITAdmin/Integrations/Slack_IT_Admin/Slack_IT_Admin.yml [hunk 14]
```diff
@@ -0,0 +1,363 @@
...
+    outputs:
+    - contextPath: Slack.Group.id
+      description: Group Id
+      type: Unknown
+    - contextPath: Slack.Group.displayName
+      description: Display Name of the Group
+      type: Unknown
+    - contextPath: Slack.Group.members
+      description: Id of the group member
+      type: Unknown
```
[the remark this answers is not in the document]
Comment (sonaaru) [reply Packs/SlackITAdmin/Integrations/Slack_IT_Admin/Slack_IT_Admin.yml]: Updated.

File: Packs/SlackITAdmin/Integrations/Slack_IT_Admin/Slack_IT_Admin.py [hunk 15]
```diff
@@ -0,0 +1,724 @@
...
+        res_json
+    )
+
+
+def delete_group_command(client, args):
+    group_id = args.get('groupId')
+    group_name = args.get('groupName')
+
+    if not (group_id or group_name):
+        return_error("You must supply either 'groupId' or 'groupName")
```
[the remark this answers is not in the document]
Comment (sonaaru) [reply Packs/SlackITAdmin/Integrations/Slack_IT_Admin/Slack_IT_Admin.py]: Updated

File: Packs/SlackITAdmin/Integrations/Slack_IT_Admin/Slack_IT_Admin.py (hunk 1 above)
Comment (sonaaru) [reply Packs/SlackITAdmin/Integrations/Slack_IT_Admin/Slack_IT_Admin.py]: updated

File: Packs/SlackITAdmin/Integrations/Slack_IT_Admin/Slack_IT_Admin.py (hunk 2 above)
Comment (sonaaru) [reply Packs/SlackITAdmin/Integrations/Slack_IT_Admin/Slack_IT_Admin.py]: updated

File: Packs/SlackITAdmin/Integrations/Slack_IT_Admin/Slack_IT_Admin.yml [hunk 16]
```diff
@@ -0,0 +1,363 @@
...
+      description: Account information
+      type: Unknown
+  dockerimage: demisto/python3:3.8.5.10845
+  runonce: false
+  script: ''
+  subtype: python3
+  type: python
+fromversion: 6.0.0
+tests:
+- No tests (auto formatted)
```
[the remark this answers is not in the document]
Comment (sonaaru) [reply Packs/SlackITAdmin/Integrations/Slack_IT_Admin/Slack_IT_Admin.yml:363]: Yes. Please check with Rashmi for credentials.

File: Packs/SlackITAdmin/Integrations/Slack_IT_Admin/Slack_IT_Admin.py (hunk 3 above)
Comment (sonaaru) [reply Packs/SlackITAdmin/Integrations/Slack_IT_Admin/Slack_IT_Admin.py]: Removed unwanted Keys.

File: Packs/SlackITAdmin/Integrations/Slack_IT_Admin/Slack_IT_Admin.yml [hunk 17]
```diff
@@ -0,0 +1,363 @@
...
+    - contextPath: CreateUser.instanceName
+      description: Name of the instance used for testing
+      type: string
+    - contextPath: CreateUser.success
+      description: Status of the result. Can be true or false
+      type: boolean
+    - contextPath: CreateUser.username
+      description: Value of username passed as argument
+      type: string
+    - contextPath: Account
```
[the remark this answers is not in the document]
Comment (sonaaru) [reply Packs/SlackITAdmin/Integrations/Slack_IT_Admin/Slack_IT_Admin.yml]: Removed this output

**Review by dantavori — APPROVED** [review]

**Pull request merged**